src/ORM/Query.php

```diff
@@ -264,6 +318,408 @@ public function selectAllExcept(Table|Association $table, array $excludedFields,
+     * to be selected, unless the second argument is set to true.
+     *
+     * ### Examples:
+     *
+     * ```
+     * $query->order(['title' => 'DESC', 'author_id' => 'ASC']);
+     * ```
+     *
+     * Produces:
+     *
+     * `ORDER BY title DESC, author_id ASC`
+     *
+     * ```
+     * $query
+     *     ->order(['title' => $query->newExpr('DESC NULLS FIRST')])
+     *     ->order('author_id');
+     * ```
+     *
+     * Will generate:
+     *
+     * `ORDER BY title DESC NULLS FIRST, author_id`
+     *
+     * ```
+     * $expression = $query->newExpr()->add(['id % 2 = 0']);
+     * $query->order($expression)->order(['title' => 'ASC']);
+     * ```
+     *
+     * and
+     *
+     * ```
+     * $query->order(function ($exp, $query) {
+     *     return [$exp->add(['id % 2 = 0']), 'title' => 'ASC'];
+     * });
+     * ```
+     *
+     * Will both become:
+     *
+     * `ORDER BY (id %2 = 0), title ASC`
+     *
+     * Order fields/directions are not sanitized by the query builder.
+     * You should use an allowed list of fields/directions when passing
+     * in user-supplied data to `order()`.
+     *
+     * If you need to set complex expressions as order conditions, you
+     * should use `orderAsc()` or `orderDesc()`.
+     *
+     * @param \Cake\Database\ExpressionInterface|\Closure|array|string $fields fields to be added to the list
+     * @param bool $overwrite whether to reset order with field list or not
+     * @return $this
+     */
+    public function order(ExpressionInterface|Closure|array|string $fields, bool $overwrite = false)
+    {
+        $this->dbQuery()->order($fields, $overwrite);
+
+        return $this;
+    }
+
+    /**
+     * Adds a complete query to be used in conjunction with an UNION operator with
+     * this query. This is used to combine the result set of this query with the one
+     * that will be returned by the passed query. You can add as many queries as you
+     * required by calling multiple times this method with different queries.
+     *
+     * By default, the UNION operator will remove duplicate rows, if you wish to include
+     * every row for all queries, use unionAll().
+     *
+     * ### Examples
+     *
+     * ```
+     * $union = (new SelectQuery($conn))->select(['id', 'title'])->from(['a' => 'articles']);
+     * $query->select(['id', 'name'])->from(['d' => 'things'])->union($union);
+     * ```
+     *
+     * Will produce:
+     *
+     * `SELECT id, name FROM things d UNION SELECT id, title FROM articles a`
+     *
+     * @param \Cake\Database\Query\SelectQuery|\Cake\ORM\Query|string $query full SQL query to be used in UNION operator
+     * @param bool $overwrite whether to reset the list of queries to be operated or not
+     * @return $this
+     */
+    public function union(SelectQuery|Query|string $query, bool $overwrite = false)
+    {
+        if ($query instanceof Query) {
+            $query = $query->sql();
```

Comment:
The query needs to be converted to string before hand as the 1st argument of `SelectQuery::union()` is typed as `Database\Query|string` so can't pass it `ORM\Query` instance.

Reply:
Wouldn't this make `$query` use a separate value binder, and therefore cause the possible bound values to be lost, and possibly create a clash with placeholder names and counts?

Reply:
Umm.. `$query` is already a separate instance. So it's value binder would be separate from the value binder for `$his`.

Reply:
Exactly, and that would be a problem, wouldn't it?

Previously both query objects would be compiled by the same query compiler, hence they shared the same value binder instance, now the passed in query uses its own binder, which neither the binder not the compiler for the `$this` query know about, hence the values that the `$query` binder holds would get lost, and subsequently the separate binder for the `$this` query could then also produce duplicate placeholder names.

Reply:
Umm right, the compiler passes the the current queries' binder to the union query too.

https://github.com/cakephp/cakephp/blob/611b274200402db07b315d0e56ff6ff1583c6066/src/Database/QueryCompiler.php#L351-L354

But I don't understand why the binder needs to be shared. As shown in the e.g. code in docblock `$union` is a separate instance from `$query`. So `$union` already has it's own value binder instance.

 ```
$union = (new Query($conn))->select(['id', 'title'])->from(['a' => 'articles']);
$query->select(['id', 'name'])->from(['d' => 'things'])->union($union);
```

Reply:
It needs to be shared because otherwise the values of `$queryB` are being lost, as `sql()` will not compile the values into the generated SQL, they will sit in the  `$queryB`'s value binder, which is never used.

Let me show an example for use in `tests/TestCase/ORM/QueryTest.php`:

```php
public function testUnionBindings()
{
    $table = $this->getTableLocator()->get('Articles');

    $queryA = $table->find()->select('id')->where(['published' => 'N']);
    $queryB = $table->find()->select('id')->where(['published' => 'Y', 'author_id !=' => 3]);

    $union = $queryA->union($queryB);

    debug($union->sql());
    debug($union->innerQuery()->getValueBinder()->bindings());

    $union->all();
}
```

This will produce the following SQL, where you can see the duplicate placeholder `:c0`, because the binder for `$union`/`$queryA` doesn't know about the ones from `$queryB`:

```none
CORE\tests\TestCase\ORM\QueryTest.php (line 3990)
########## DEBUG ##########
'(SELECT Articles.id AS Articles__id FROM articles Articles WHERE published = :c0)
UNION (SELECT Articles.id AS Articles__id FROM articles Articles WHERE (published = :c0 AND author_id != :c1))'
###########################
```

Subsequently the only value binder that is actually being used when the final SQL is being compiled, will only hold the values from `$queryA`:

```none
CORE\tests\TestCase\ORM\QueryTest.php (line 3991)
########## DEBUG ##########
[
  ':c0' => [
    'value' => 'N',
    'type' => 'string',
    'placeholder' => 'c0'
  ]
]
###########################
```` 

and lastly a DBMS that requires the parameter and placeholder count to match, will throw an error:

```none
PDOException : SQLSTATE[HY093]: Invalid parameter number: number of bound variables does not match number of tokens
 src/Database/Statement/Statement.php:138
 src/Database/Driver.php:277
 src/Database/Driver.php:262
 src/Database/Connection.php:280
 src/Core/Retry/CommandRetry.php:71
 src/Database/Connection.php:280
 src/Database/Query.php:205
 src/Database/Query/SelectQuery.php:108
 src/ORM/Query.php:1617
 src/Datasource/QueryTrait.php:290
 src/ORM/Query.php:1538
 tests/TestCase/ORM/QueryTest.php:3993
```

Reply:
Thank you, I'll add this test case and try to figure out a solution.

Reply:
@ndm2 I resolved the issue by adding a new `Cake\Database\QueryInterface` and using that for typing `SelectQuery::union()`. Using the new interface instead of the concrete `ORM\Query` at various other places helped fixes lot of other failures too.

Reply:
I see. Any plans to further populate the interface, or is this just meant for the purpose where the query is only ever being compiled?

Reply:
Not really, just needed a way to indicate that the object represents a full query, since a various places during query complication a full query needs to be wrapped in parenthesis.